core/domain/wipeout_domain.py

```diff
@@ -67,4 +68,15 @@ def create_default(cls, user_id, email, exploration_ids, collection_ids):
             PendingDeletionRequest. The default pending deletion request
             domain object.
         """
-        return cls(user_id, email, False, exploration_ids, collection_ids, None)
+        return cls(user_id, email, False, exploration_ids, collection_ids, {})
+
+    def validate(self):
+        """Checks that the domain object is valid.
+
+        Raises:
+            ValidationError: activity_mappings contain wrong key.
+        """
+        for key in self.activity_mappings.keys():
```

Comment:
Actually, one question. Shouldn't the keys of activity_mappings be a fixed set? I.e. we should still include keys even if there is no entity of that type?

Reply:
I don't think so, I think that when the key is not yet included it means that we didn't yet even started processing that kind of models and is useful for the wipeout procedure.

We use `if activity_category not in pending_deletion_request.activity_mappings` meaning that we need to generated mappings for this category. If there is `{activity_category: {}}` it means that the mappings were generated but that the user doesn't have any `activity_category` model.

Reply:
Oh, I see what you mean. Two follow-up thoughts:

- Could you please add documentation to the storage model field to explain this? It is not usual for us to have dicts with variable key sets so I think the difference between {} and the key not being there needs some explanation. Add a similar explanation above the `if activity_category not in pending_deletion_request.activity_mappings` line you quoted, to explain why that case may occur.
- Separately: should we modify the release process to ensure that PendingDeletionModel is completely empty before initiating a release? I can see issues if a new release does a migration of data and introduces new fields.

Reply:
Here's a suggested docstring template for the storage layer:

```
Each key of this dict should be one of .... The corresponding values are dicts whose key-value pairs each represent ... If a key is absent from the activity_mappings dict, this means that ...
```

and for the domain layer:

```
The pending deletion request model might have only been partially generated, so we fill in the missing fields here.
```

Feel free to modify as needed.

Reply:
> Could you please add documentation to the storage model field to explain this? It is not usual for us to have dicts with variable key sets so I think the difference between {} and the key not being there needs some explanation. Add a similar explanation above the if activity_category not in pending_deletion_request.activity_mappings line you quoted, to explain why that case may occur.

Done.

> If it's possible for mappings to not be generated for a particular category due to things changing in the interim, what prevents the mappings that have already been generated from becoming wrong over time? How can we mitigate that sort of thing (if applicable)?

For the activity categories that are implemented this is not a problem since new models can only be created by the user that requested deletion (who is blocked from accessing the page). I do not think this will be a problem for other activity categories that I have to yet implement I will make sure that this is not a problem.

> Separately: should we modify the release process to ensure that PendingDeletionModel is completely empty before initiating a release? I can see issues if a new release does a migration of data and introduces new fields.
Hmm, this makes sense, all the deletion requests should always be completed in the same release version in which they were requested.